src/agreementManager/AgreementManager.ts

```diff
@@ -77,45 +111,191 @@ class AgreementManager {
+
+    public getJoinChannelChain(): JoinChannelBlockStruct[] {
+        const blocks: JoinChannelBlockStruct[] = [];
+        const channelBlocks = this.joinChannelBlocks;
+
+        if (!channelBlocks) return blocks;
+
+        let currentHash = this.latestJoinChannelBlockHash;
+        if (!currentHash) return blocks;
+
+        while (true) {
+            const block = channelBlocks.get(currentHash);
+            if (!block) {
+                throw new Error(`Chain is broken at hash ${currentHash}`);
+            }
+
+            blocks.unshift(block);
+
+            // Check if this is the genesis block (block's hash equals its previousBlockHash)
+            const blockHash = ethers.keccak256(
+                EvmUtils.encodeJoinChannelBlock(block)
+            );
+
+            if (blockHash === block.previousBlockHash) {
+                // We've reached the genesis block
+                break;
+            }
 
-        agreement.blockSignatures.push(confirmationSignature);
+            currentHash = block.previousBlockHash;
+        }
+
+        return blocks;
+    }
+
+    public getLatestJoinChannelBlockHash(): BytesLike {
+        return this.latestJoinChannelBlockHash;
     }
+
+    public getExitChannelChain(): ExitChannelBlockStruct[] {
+        const blocks: ExitChannelBlockStruct[] = [];
+        const channelBlocks = this.exitChannelBlocks;
+
+        if (!channelBlocks) return blocks;
+
+        let currentHash = this.latestExitChannelBlockHash;
+        if (!currentHash) return blocks;
+
+        while (true) {
+            const block = channelBlocks.get(currentHash);
+            if (!block) {
+                throw new Error(`Chain is broken at hash ${currentHash}`);
+            }
+
+            blocks.unshift(block);
+
+            // Check if this is the genesis block (block's hash equals its previousBlockHash)
+            const blockHash = ethers.keccak256(
+                EvmUtils.encodeExitChannelBlock(block)
+            );
+
+            if (blockHash === block.previousBlockHash) {
+                // We've reached the genesis block
+                break;
+            }
+
+            currentHash = block.previousBlockHash;
+        }
+
+        return blocks;
+    }
+
+    public getLatestExitChannelBlockHash(): BytesLike {
+        return this.latestExitChannelBlockHash;
+    }
+
+    public getStateMachineState(
+        forkCnt: number,
+        transactionCnt: number
+    ): BytesLike | undefined {
+        return this.forkService.getAgreement(forkCnt, transactionCnt)
+            ?.encodedState;
+    }
+
+    public getForkProofSignedBlocks(forkCnt: number): SignedBlockStruct[] {
+        const lastMilestoneBlock = EvmUtils.decodeBlock(
+            this.forkService.getForkProof(forkCnt)!.forkMilestoneProofs[-1]
+                .blockConfirmations[-1].signedBlock.encodedBlock
+        );
+        const lastMilestoneBlockHeight = lastMilestoneBlock.transaction.header
+            .transactionCnt as number;
```

Comment:
type casting isn't casting - tricking typescript to think it's a number when it's not doesn't help - to transform it to a number you use Number(...)